File: reverse_linked_list.go
func reverseList(head *ListNode) *ListNode {
    var prev, next *ListNode

    for head != nil {
        next = head.Next
        head.Next = prev
        prev = head
        head = next
    }
    return prev
}

// Time Complexity : O(n)
// Space Complexity : O(1)

File: reverse_linked_list.cpp
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *next = nullptr, *prev = nullptr;
        
        while(head) {
            next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }
};

// Time Complexity : O(n)
// Space Complexity : O(1)